Modules/System/AL Rest Client/src/ALRestClientImpl.Codeunit.al

```diff
@@ -0,0 +1,153 @@
+
+    procedure Initialize()
+    begin
+        ClearAll();
+
+        HttpClient.Clear();
+        HttpClient.Timeout := 60000;
+        Initialized := true;
+        SetDefaultUserAgentHeader();
+    end;
+
+    procedure SetBaseAddress(Url: Text)
+    begin
+        AssertInitialized();
+        HttpClient.SetBaseAddress(Url);
+    end;
+
+    procedure GetBaseAddress() Url: Text
+    begin
+        AssertInitialized();
+        Url := HttpClient.GetBaseAddress;
+    end;
+
+    procedure SetTimeOut(TimeOut: Duration)
+    begin
+        AssertInitialized();
+        HttpClient.Timeout := TimeOut;
+    end;
+
+    [NonDebuggable]
+    procedure SetAuthorizationHeader(Value: Text)
+    begin
+        SetDefaultRequestHeader('Authorization', Value);
+    end;
+
+    procedure SetUserAgentHeader(Value: Text)
+    begin
+        SetDefaultRequestHeader('User-Agent', Value);
+    end;
+
+    local procedure SetDefaultUserAgentHeader()
+    begin
+        SetUserAgentHeader(GetUserAgentString());
+    end;
+
+    local procedure GetUserAgentString() UserAgentString: Text
+    var
+        ModuleInfo: ModuleInfo;
+    begin
+        if NavApp.GetCurrentModuleInfo(ModuleInfo) then
+            UserAgentString := StrSubstNo(UserAgentLbl, ModuleInfo.Publisher(), ModuleInfo.Name(), ModuleInfo.AppVersion());
+    end;
+
+    procedure SetDefaultRequestHeader(Name: Text; Value: Text)
+    begin
+        AssertInitialized();
+        if HttpClient.DefaultRequestHeaders.Contains(Name) then
+            HttpClient.DefaultRequestHeaders.Remove(Name);
+        HttpClient.DefaultRequestHeaders.Add(Name, Value);
+    end;
+
+    [NonDebuggable]
+    procedure SetAuthentication(HttpAuthenticationInstance: Interface "Http Authentication")
+    begin
+        HttpAuthentication := HttpAuthenticationInstance;
+    end;
+
+    procedure HasConnectionError(): Boolean
+    begin
+        exit(ConnectionError);
+    end;
+
+    local procedure AssertInitialized()
+    begin
+        if not Initialized then
+            Error(NotInitializedErr);
+    end;
+
+    procedure SendRequest(var ALHttpRequestMessage: Codeunit "AL Http Request Message") ALHttpResponseMessage: Codeunit "AL Http Response Message"
+    begin
+        if not TrySendRequest(ALHttpRequestMessage, ALHttpResponseMessage) then;
+    end;
+
+    [TryFunction]
+    local procedure TrySendRequest(var ALHttpRequestMessage: Codeunit "AL Http Request Message"; var ALHttpResponseMessage: Codeunit "AL Http Response Message")
+    var
+        ErrorMessage: Text;
+        HttpRequestMessage: HttpRequestMessage;
+        HttpResponseMessage: HttpResponseMessage;
+    begin
+        AssertInitialized();
+
+        ALHttpResponseMessage.Initialize();
+        ConnectionError := false;
+
+        if HttpAuthentication.IsAuthenticationRequired() then
+            if not Authorize(ALHttpRequestMessage) then begin
+                ErrorMessage := GetLastErrorText();
```

Comment:
It would be nice to send some telemetry when we encounter errors. We should of course make sure that the error text could never have customer content (e. g. information from the request).
This applies to the errors below in this procedure as well.

Reply:
The code has been redesigned to not handle an error from authorization. Instead, the error will just be raised. It is a runtime error anyway, and there is no need to handle it differently than any other error.
The other error conditions are also not raising an error anymore. If the connection fails, either because of outgoing calls being blocked or not receiving a response, then the SendRequest method returns false and an error is raised from the public Rest Client codeunit.
The error message for unsuccessful HTTP status is only used from the Json methods of the public Rest Client codeunit.
After all, this is in line with the default behavior of the HttpClient.

Reply:
The methods of the public Rest Client codeunit could send telemetry. But it would only end up in the customers telemetry and not in the ISV's telemetry (apart from Microsoft).
This is up for discussion, please advise.

Reply:
Let's address this in a followup PR. We'll go with this PR as the initial version.